### PR TITLE
[TritonIntelGPUToLLVM] Detect more sub-group shuffle `convert_layout`

### PR DESCRIPTION
**Pull request by victor-eds**
Detect sub-group shuffle `convert_layout` cases of more than one element per thread.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM.cpp [hunk 1]
```diff
@@ -477,35 +528,13 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
...
-               {kLane, buildBasis(laneSize, laneIndex)}}};
-    std::array<StringAttr, 2> outDimNames{kRegister, kLane};
-    return conversion == LinearLayout(bases, outDimNames);
+    int32_t registerInDimSize = conversion->getInDimSize(kRegister);
+    int32_t laneInDimSize = conversion->getInDimSize(kLane);
+    return conversion->getBases().lookup(kRegister) ==
+               buildSubGroupTransposeRegisterBases(registerInDimSize,
+                                                   laneInDimSize) &&
+           conversion->getBases().lookup(kLane) ==
+               buildSubGroupTransposeLaneBases(laneInDimSize);
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM.cpp:537]: Refactor and protected against some illegal layout creation.

File: third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM.cpp [hunk 2]
```diff
@@ -612,39 +641,29 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
 
     LinearLayout comp = dstLayout.invertAndCompose(srcLayout);
     std::optional<LinearLayout> conversion = comp.divideRight(
+        LinearLayout::zeros1D(comp.getInDimSize(kLane), kLane, kLane) *
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM.cpp:644]: Should always be 0 for this case

File: third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM.cpp [hunk 3]
```diff
@@ -612,39 +641,29 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
...
         LinearLayout::identity1D(comp.getInDimSize(kWarp), kWarp, kWarp) *
         LinearLayout::identity1D(comp.getInDimSize(kBlock), kBlock, kBlock));
-    assert(conversion && "Expecting valid conversion");
+    if (!conversion)
+      return false;
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM.cpp:648]: May be `None` is 0 check fails.

File: third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM.cpp [hunk 4]
```diff
@@ -619,32 +653,27 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
...
+    //   lane=2**M -> (0, 0)
+    // where out dims are: [register (size 2**(N - M)), lane (size 2**(M + 1))]
+    //
+    // With N >= M.
+    int32_t registerInDimSize = conversion->getInDimSize(kRegister);
+    int32_t laneOutDimSize = conversion->getOutDimSize(kLane);
+    return conversion->sublayoutIsZero({kLane}, {kRegister, kLane}) &&
+           conversion->getBases().lookup(kRegister) ==
+               buildSubGroupShuffleRegisterBases(registerInDimSize,
+                                                 laneOutDimSize);
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM.cpp:676]: I will investigate whether there is a better way to express this in the future when generalizing. Same for transpose case.

File: test/Conversion/intel/sub-group-shuffle.mlir [hunk 5]
```diff
@@ -257,3 +257,106 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 1 :
...
+  // CHECK:           llvm.call spir_funccc @_Z17sub_group_shuffledj(%[[VAL_3]], %[[VAL_86]])
+  // CHECK:           %[[VAL_89:.*]] = llvm.mlir.constant(12 : i32) : i32
+  // CHECK:           llvm.call spir_funccc @_Z17sub_group_shuffledj(%[[VAL_3]], %[[VAL_89]])
+  // CHECK:           %[[VAL_92:.*]] = llvm.mlir.constant(13 : i32) : i32
+  // CHECK:           llvm.call spir_funccc @_Z17sub_group_shuffledj(%[[VAL_3]], %[[VAL_92]])
+  // CHECK:           %[[VAL_95:.*]] = llvm.mlir.constant(14 : i32) : i32
+  // CHECK:           llvm.call spir_funccc @_Z17sub_group_shuffledj(%[[VAL_3]], %[[VAL_95]])
+  // CHECK:           %[[VAL_98:.*]] = llvm.mlir.constant(15 : i32) : i32
+  // CHECK:           llvm.call spir_funccc @_Z17sub_group_shuffledj(%[[VAL_3]], %[[VAL_98]])
+  tt.func @test(%arg0: tensor<32xf64, #sliced>) -> tensor<32xf64, #sliced1> {
```
Comment (mfrancepillois) [inline test/Conversion/intel/sub-group-shuffle.mlir:340]: Can you use a more explicit name?
```suggestion
  tt.func @test_f64(%arg0: tensor<32xf64, #sliced>) -> tensor<32xf64, #sliced1> {
```

File: test/Conversion/intel/sub-group-shuffle.mlir [hunk 6]
```diff
@@ -257,3 +257,106 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 1 :
...
+
+#blocked = #triton_gpu.blocked<{sizePerThread = [1, 16], threadsPerWarp = [16, 1], warpsPerCTA = [2, 1], order = [0, 1]}>
+#blocked1 = #triton_gpu.blocked<{sizePerThread = [16, 1], threadsPerWarp = [1, 16], warpsPerCTA = [2, 1], order = [0, 1]}>
+#sliced = #triton_gpu.slice<{dim = 1, parent = #blocked}>
+#sliced1 = #triton_gpu.slice<{dim = 1, parent = #blocked1}>
+
+module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 2 : i32, "triton_gpu.threads-per-warp" = 16 : i32} {
+  // CHECK-LABEL:     llvm.func spir_kernelcc @test
+  // CHECK-COUNT-64:    llvm.call spir_funccc @_Z17sub_group_shuffleij
+  tt.func @test(%arg0: tensor<128xi32, #sliced>) -> tensor<128xi32, #sliced1> {
```
Comment (mfrancepillois) [inline test/Conversion/intel/sub-group-shuffle.mlir:358]: Same as above.
```suggestion
  tt.func @test_i32_2warps(%arg0: tensor<128xi32, #sliced>) -> tensor<128xi32, #sliced1> {
```

**Review by etiotto — APPROVED** [review]

**Review by mfrancepillois — APPROVED** [review]
I'm missing some background on subgroup shuffle, but PR LGTM.

**Pull request merged**